src/generator/qbe.rs

```diff
@@ -0,0 +1,1130 @@
+                    // TODO: Cast to function return type
+                    func.add_instr(QbeInstr::Ret(Some(result)));
+                }
+                None => func.add_instr(QbeInstr::Ret(None)),
+            },
+            Statement::If(cond, if_clause, else_clause) => {
+                self.generate_if(func, cond, if_clause, else_clause)?;
+            }
+            Statement::While(cond, body) => {
+                self.generate_while(func, cond, body)?;
+            }
+            Statement::Break => {
+                if let Some(label) = &self.loop_label {
+                    func.add_instr(QbeInstr::Jmp(format!("{}.end", label)));
+                } else {
+                    return Err("break used outside of a loop".to_owned());
+                }
+            }
+            Statement::Continue => {
+                if let Some(label) = &self.loop_label {
+                    func.add_instr(QbeInstr::Jmp(format!("{}.cond", label)));
+                } else {
+                    return Err("continue used outside of a loop".to_owned());
+                }
+            }
+            Statement::Exp(expr) => {
+                self.generate_expression(func, expr)?;
+            }
+            _ => todo!("statement: {:?}", stmt),
+        }
+        Ok(())
+    }
+
+    /// Generates an expression
+    fn generate_expression(
+        &mut self,
+        func: &mut QbeFunction,
+        expr: &Expression,
+    ) -> GeneratorResult<(QbeType, QbeValue)> {
+        match expr {
+            Expression::Int(literal) => {
+                let tmp = self.new_temporary();
+                func.assign_instr(
+                    tmp.clone(),
+                    QbeType::Word,
+                    QbeInstr::Copy(QbeValue::Const(*literal as u64)),
+                );
+
+                Ok((QbeType::Word, tmp))
+            }
+            Expression::Str(string) => self.generate_string(string),
+            Expression::Bool(literal) => {
+                let tmp = self.new_temporary();
+                func.assign_instr(
+                    tmp.clone(),
+                    QbeType::Word,
+                    QbeInstr::Copy(QbeValue::Const(if *literal { 1 } else { 0 })),
+                );
+
+                Ok((QbeType::Word, tmp))
+            }
+            Expression::FunctionCall(name, args) => {
+                let mut new_args: Vec<(QbeType, QbeValue)> = Vec::new();
+                for arg in args.iter() {
+                    new_args.push(self.generate_expression(func, &arg)?);
+                }
+
+                let tmp = self.new_temporary();
+                func.assign_instr(
+                    tmp.clone(),
+                    // TODO: get that type properly
+                    QbeType::Word,
+                    QbeInstr::Call(name.clone(), new_args),
+                );
+
+                Ok((QbeType::Word, tmp))
+            }
+            Expression::Variable(name) => self.get_var(name).map(|v| v.to_owned()),
```

Comment:
Does this compile? I can't test it rn.

```suggestion
            Expression::Variable(name) => self.get_var(name).map(to_owned),
```

Reply:
No, because `to_owned` is not in the current scope. It will probably work with `ToOwned::to_owned`, but the current variant is more concise to me